"Hooks FAQ" - "Хуки - вопросы и ответы" - ready (#208)
reactjs/ru.reactjs.org: content/docs/hooks-faq.md

---
id: hooks-faq
title: Хуки - вопросы и ответы
permalink: docs/hooks-faq.html
prev: hooks-reference.html
---

> *Хуки* добавлены в React 16.8. С их помощью вы можете использовать состояние и другие возможности React без необходимости использовать классы.

Reviewer: Тут #94 происходило обсуждение перевода шапки про хуки и в ишью можно найти конечный перевод, который используется на всех страницах.
```suggestion
*Хуки* — нововведение в React 16.8, которое позволяет использовать состояние и другие возможности React без написания классов.
```

> *Хуки* были добавлены в React 16.8. С их помощью вы можете использовать состояние и другие возможность React без необходимости использовать классы.

Reviewer 2: ```suggestion
*Хуки* были добавлены в React 16.8. С их помощью вы можете использовать состояние и другие возможности React без необходимости использовать классы.
```

  ).join('\n')
-->

* **[Внедрение хуков](#adoption-strategy)**
  * [В какой версии React появились хуки?](#which-versions-of-react-include-hooks)
  * [Надо ли переписать все мои классовые компоненты?](#do-i-need-to-rewrite-all-my-class-components)
  * [Что можно сделать с помощью хуков, что невозможно с помощью классов?](#what-can-i-do-with-hooks-that-i-couldnt-with-classes)

>   * [Какая часть моих знаний о React всё еще актуальна?](#how-much-of-my-react-knowledge-stays-relevant)

Reviewer: ```suggestion
  * [Какая часть моих знаний о React всё ещё актуальна?](#how-much-of-my-react-knowledge-stays-relevant)
```

  * [Что мне использовать: хуки, классы или оба подхода?](#should-i-use-hooks-classes-or-a-mix-of-both)
  * [Дают ли хуки все возможности классов?](#do-hooks-cover-all-use-cases-for-classes)
  * [Являются ли хуки заменой рендер-пропсам и компонентам высшего порядка?](#do-hooks-replace-render-props-and-higher-order-components)

>   * [Как хуки повлияют на популярные API - например Redux `connect()` и React Router?](#what-do-hooks-mean-for-popular-apis-like-redux-connect-and-react-router)

Reviewer: ```suggestion
  * [Как хуки повлияют на популярные API – например Redux `connect()` и React Router?](#what-do-hooks-mean-for-popular-apis-like-redux-connect-and-react-router)
```

  * [Поддерживают ли хуки статическую типизацию?](#do-hooks-work-with-static-typing)
  * [Как тестировать компоненты, которые используют хуки?](#how-to-test-components-that-use-hooks)

>   * [Что конкретно правила линтера проверяют в хуках?](#what-exactly-do-the-lint-rules-enforce)

Reviewer: Мне кажется, что такой порядок слов лучше на слух будет. Как вы думаете?
```suggestion
  * [Что конкретно проверяют правила линтера в хуках?](#what-exactly-do-the-lint-rules-enforce)
```

* **[От классов к хукам](#from-classes-to-hooks)**
  * [Как методы жизненного цикла соответствуют хукам?](#how-do-lifecycle-methods-correspond-to-hooks)
  * [Существует что-нибудь наподобие полей экземпляра?](#is-there-something-like-instance-variables)

>   * [Сколько переменных состояния я могу использовать - одну или несколько?](#should-i-use-one-or-many-state-variables)

Reviewer: ```suggestion
  * [Сколько переменных состояния я могу использовать – одну или несколько?](#should-i-use-one-or-many-state-variables)
```

  * [Могу ли я использовать эффект только на обновлениях компонента?](#can-i-run-an-effect-only-on-updates)
  * [Как получить прошлые пропсы или состояние](#how-to-get-the-previous-props-or-state)
  * [Как я могу реализовать getDerivedStateFromProps?](#how-do-i-implement-getderivedstatefromprops)
  * [Существует что-нибудь наподобие forceUpdate?](#is-there-something-like-forceupdate)
  * [Могу ли я изменить реф, переданный в функциональный компонент?](#can-i-make-a-ref-to-a-function-component)
  * [Что значит `const [thing, setThing] = useState()`](#what-does-const-thing-setthing--usestate-mean)
* **[Оптимизации производительности](#performance-optimizations)**
  * [Могу ли я пропустить эффект при обновлениях?](#can-i-skip-an-effect-on-updates)
  * [Как я могу реализовать shouldComponentUpdate?](#how-do-i-implement-shouldcomponentupdate)

>   * [Как закэшировать вычисления?](#how-to-memoize-calculations)

Reviewer 3: С небольшим отрывов используется чаще с буквой "е".

```suggestion
  * [Как закешировать вычисления?](#how-to-memoize-calculations)
```

  * [Как лениво создавать большие объекты?](#how-to-create-expensive-objects-lazily)
  * [Являются ли хуки медленными из-за создания функций на каждом рендере?](#are-hooks-slow-because-of-creating-functions-in-render)
  * [Как избежать передачи колбэков вниз?](#how-to-avoid-passing-callbacks-down)
  * [Как получить часто изменяемое значение из хука `useCallback`?](#how-to-read-an-often-changing-value-from-usecallback)
* **[Под капотом](#under-the-hood)**
  * [Как React связывает вызовы хуков с компонентом?](#how-does-react-associate-hook-calls-with-components)

>   * [Что послужило прообразом Хуков?](#what-is-the-prior-art-for-hooks)

Reviewer 3: ```suggestion
  * [Что послужило прообразом хуков?](#what-is-the-prior-art-for-hooks)
```

## Внедрение хуков {#adoption-strategy}

> ### В какой версии React появились хуки?{#which-versions-of-react-include-hooks}

Reviewer 3: ```suggestion
### В какой версии React появились хуки? {#which-versions-of-react-include-hooks}
```

Начиная с релиза 16.8.0, React включает в себя стабильную реализацию хуков для:

* React DOM
* React DOM Server
* React Test Renderer
* React Shallow Renderer

Обратите внимание, что **хуки будут доступны, только если все React-пакеты версии 16.8.0 или выше**. Хуки не будут работать, если вы, например, забыли обновить React DOM.

Поддержка хуков в React Native добавится в следующем стабильном релизе.

> ### Надо ли переписать все мои классовые компоненты?{#do-i-need-to-rewrite-all-my-class-components}

Reviewer 3: ```suggestion
### Надо ли переписать все мои классовые компоненты? {#do-i-need-to-rewrite-all-my-class-components}
```

Нет, мы [не собираемся](/docs/hooks-intro.html#gradual-adoption-strategy) удалять классы из React, да никто бы и не смог позволить себе такой глобальный рефакторинг. Мы советуем пробовать хуки только в новом коде.

### Что можно сделать с помощью хуков, что невозможно с помощью классов? {#what-can-i-do-with-hooks-that-i-couldnt-with-classes}

> Хуки дают новый мощный способ повторного использования кода в компонентах. ["Создание ваших хуков"](/docs/hooks-custom.html) даст представление об их возможностях. [Эта статья](https://medium.com/@dan_abramov/making-sense-of-react-hooks-fdbde8803889), написанная одним из участников команды React, описывает новые возможности, которые открывают хуки.

Reviewer: ```suggestion
Хуки дают новый мощный способ повторного использования кода в компонентах. [«Создание ваших хуков»](/docs/hooks-custom.html) даст представление об их возможностях. [Эта статья](https://medium.com/@dan_abramov/making-sense-of-react-hooks-fdbde8803889), написанная одним из участников команды React, описывает новые возможности, которые открывают хуки.
```

> Хуки дают новый мощный способ повторного использования кода в компонентах. [«Создание ваших хуков»](/docs/hooks-custom.html) даст представление об их возможностях. [Эта статья](https://medium.com/@dan_abramov/making-sense-of-react-hooks-fdbde8803889), написанная одним из участников команды React, описывает новые возможности, которые открывают хуки.

Reviewer 3: ```suggestion
Хуки дают новый мощный способ повторного использования кода в компонентах. [«Создание пользовательских хуков»](/docs/hooks-custom.html) даст представление об их возможностях. [Эта статья](https://medium.com/@dan_abramov/making-sense-of-react-hooks-fdbde8803889), написанная одним из участников команды React, описывает новые возможности, которые открывают хуки.
```

> ### Какая часть моих знаний о React всё еще актуальна? {#how-much-of-my-react-knowledge-stays-relevant}

Reviewer: ```suggestion
### Какая часть моих знаний о React всё ещё актуальна? {#how-much-of-my-react-knowledge-stays-relevant}
```

> Хуки -- это новый способ использовать возможности React, которые вы уже знаете: состояние, жизненный цикл, контекст и рефы. Хуки не изменили фундаментальную логику React, и ваши знания компонентов, пропсов и потока данных сверху-вниз остаются актуальными.

Reviewer 3: ```suggestion
Хуки -- это новый способ использовать возможности React, которые вы уже знаете: состояние, жизненный цикл, контекст и рефы. Хуки не изменили фундаментальную логику React, и ваши знания компонентов, пропсов и потока данных сверху вниз остаются актуальными.
```

Однако, хуки -- не самая простая часть React. Если вам чего-то не хватает в этой документации, [дайте нам знать](https://github.com/reactjs/reactjs.org/issues/new), и мы постараемся вам помочь.

### Что мне использовать: хуки, классы или оба подхода? {#should-i-use-hooks-classes-or-a-mix-of-both}

Когда вы будете готовы, мы рекомендуем начать использовать хуки в ваших новых компонентах. Убедитесь, что каждый в команде знаком с документацией и поддерживает вас. Мы не рекомендуем переписывать ваши существующие классы на хуки, пока нет необходимости возвращаться к определённому компоненту (например, чтобы исправить баг).

> Вы не можете использовать хуки *внутри* классового компонента, но вы можете комбинировать классы и функциональные компоненты с хуками в одном дереве. Классовый компонент или функциональный с хуками -- неважно, это всего-лишь особенности реализации. В будущем, мы ожидаем, что хуки станут главным способом написания React-компонентов.

Reviewer 3: ```suggestion
Вы не можете использовать хуки *внутри* классового компонента, но вы можете комбинировать классы и функциональные компоненты с хуками в одном дереве. Классовый компонент или функциональный с хуками -- неважно, это всего лишь особенности реализации. В будущем мы ожидаем, что хуки станут главным способом написания React-компонентов.
```

> Вы не можете использовать хуки *внутри* классового компонента, но вы можете комбинировать классы и функциональные компоненты с хуками в одном дереве. Классовый компонент или функциональный с хуками -- неважно, это всего-лишь особенности  реализации. В будущем, мы ожидаем, что хуки статун главным способом написания React-компонентов.

Reviewer: ```suggestion
Вы не можете использовать хуки *внутри* классового компонента, но вы можете комбинировать классы и функциональные компоненты с хуками в одном дереве. Классовый компонент или функциональный с хуками -- неважно, это всего-лишь особенности реализации. В будущем, мы ожидаем, что хуки станут главным способом написания React-компонентов.
```

### Дают ли хуки все возможности классов? {#do-hooks-cover-all-use-cases-for-classes}

Наша главная задача, чтобы хуки покрывали весь функционал классов в ближайшем будущем. Пока не существует хуков, реализующих методы жизненного цикла  `getSnapshotBeforeUpdate` и `componentDidCatch`, но мы планируем скоро их добавить.

> Хуки появились совсем недавно, и некоторые сторонние библиотеки могли еще не приспособиться к ним.

Reviewer: ```suggestion
Хуки появились совсем недавно, и некоторые сторонние библиотеки могли ещё не приспособиться к ним.
```

### Являются ли хуки заменой рендер-пропсам и компонентам высшего порядка? {#do-hooks-replace-render-props-and-higher-order-components}

> Обычно, рендер пропсы и компоненты высшего порядка рендерят один дочерний компонент, увеличивая вложенность, но добавляя к нему функционал. Мы думаем, что хуки - более простой способ сделать это. Оба подхода всё ещё имеют право на жизнь (например, компонент, следящий за скроллом, может иметь проп `renderItem`, или компонент-контейнер может иметь разные рендер-пропсы для различных слотов). Но в большинстве случаев, хуки должны помочь уменьшить вложенность компонентов в вашем дереве.

Reviewer 3: ```suggestion
Как правило, рендер-пропсы и компоненты высшего порядка рендерят один дочерний компонент, увеличивая вложенность, но добавляя к нему функционал. Мы думаем, что хуки — более простой способ сделать это. Оба подхода всё ещё имеют право на жизнь (например, компонент, следящий за скроллом, может иметь проп `renderItem`, или компонент-контейнер может иметь разные рендер-пропсы для различных слотов). Но в большинстве случаев, хуки должны помочь уменьшить вложенность компонентов в вашем дереве.
```

> Обычно, рендер пропсы и компоненты высшего порядка рендерят один дочерний компонент, увеличивая вложенность, но добавляя к нему функционал. Мы думаем, что Хуки - более простой способ сделать это. Оба подхода всё еще имеют право на жизнь (например, компонент, следяющий за скроллом, может иметь проп `renderItem`, или компонент-контейнер может иметь разные рендер-пропсы для различных слотов). Но в большинстве случаев, хуки должны помочь уменьшить вложенность компонентов в вашем дереве.

Reviewer: ```suggestion
Обычно, рендер пропсы и компоненты высшего порядка рендерят один дочерний компонент, увеличивая вложенность, но добавляя к нему функционал. Мы думаем, что хуки - более простой способ сделать это. Оба подхода всё ещё имеют право на жизнь (например, компонент, следящий за скроллом, может иметь проп `renderItem`, или компонент-контейнер может иметь разные рендер-пропсы для различных слотов). Но в большинстве случаев, хуки должны помочь уменьшить вложенность компонентов в вашем дереве.
```

> ### Как Хуки повлияют на популярные API – например Redux `connect()` и React Router? {#what-do-hooks-mean-for-popular-apis-like-redux-connect-and-react-router}

Reviewer 3: ```suggestion
### Как хуки повлияют на популярные API – например Redux `connect()` и React Router? {#what-do-hooks-mean-for-popular-apis-like-redux-connect-and-react-router}
```

> ### Как Хуки повлияют на популярные API - например Redux `connect()` и React Router? {#what-do-hooks-mean-for-popular-apis-like-redux-connect-and-react-router}

Reviewer: ```suggestion
### Как Хуки повлияют на популярные API – например Redux `connect()` и React Router? {#what-do-hooks-mean-for-popular-apis-like-redux-connect-and-react-router}
```

> Используйте тот же API, который вы использовали до этого - всё продолжает работать.

Reviewer: ```suggestion
Используйте тот же API, который вы использовали до этого – всё продолжает работать.
```

> В будущем, новые версии этих библиотек могут добавить специальные хуки, например `useRedux()` или  `useRouter()`, которые позволят вам использовать тот же функционал без необходимости оборачивать компоненты.

Reviewer 3: ```suggestion
В будущем, новые версии этих библиотек могут добавить специальные хуки, например `useRedux()` или `useRouter()`, которые позволят вам использовать ту же функциональность без необходимости оборачивать компоненты.
```

### Поддерживают ли хуки статическую типизацию? {#do-hooks-work-with-static-typing}

Хуки спроектированы с учётом статической типизации. Так как они являются функциями, их легче типизировать правильным образом (например, в сравнении с компонентами высшего порядка). Последние версии Flow и TypeScript включают в себя поддержку React-хуков.

Importantly, custom Hooks give you the power to constrain React API if you'd like to type them more strictly in some way. React gives you the primitives, but you can combine them in different ways than what we provide out of the box. TODO

> ### Как тестировать компоненты, которые используют Хуки? {#how-to-test-components-that-use-hooks}

Reviewer 3: ```suggestion
### Как тестировать компоненты, которые используют хуки? {#how-to-test-components-that-use-hooks}
```

С точки зрения React, компонент, использующий хуки, является обычным компонентом. Если ваш способ тестирования не опирается на внутреннюю часть React, тестирование компонентов с хуками не должно отличатся от тестирования других компонентов.

Например, у нас есть следующий компонент-счетчик:

```js
function Example() {
  const [count, setCount] = useState(0);
  useEffect(() => {
    document.title = `Вы нажали ${count} раз`;
  });
  return (
    <div>
      <p>Вы нажали {count} раз</p>
      <button onClick={() => setCount(count + 1)}>

>         Нажмите на меня

Reviewer 3: ```suggestion
        Нажми на меня
```

  ).join('\n')
-->

* **[Внедрение хуков](#adoption-strategy)**
  * [В какой версии React появились хуки?](#which-versions-of-react-include-hooks)
  * [Надо ли переписать все мои классовые компоненты?](#do-i-need-to-rewrite-all-my-class-components)
  * [Что я могу сделать с помощью Хуков, что не смог бы с помощью классов?](#what-can-i-do-with-hooks-that-i-couldnt-with-classes)
  * [Какая часть моих знаний о React всё еще актуальна?](#how-much-of-my-react-knowledge-stays-relevant)
  * [Что мне использовать: Хуки, классы или оба подхода?](#should-i-use-hooks-classes-or-a-mix-of-both)
  * [Дают ли Хуки все возможности классов?](#do-hooks-cover-all-use-cases-for-classes)
  * [Являются ли Хуки заменой рендер-пропсам и компонентам высшего порядка?](#do-hooks-replace-render-props-and-higher-order-components)
  * [Как Хуки повлияют на популярные API - например Redux `connect()` и React Router?](#what-do-hooks-mean-for-popular-apis-like-redux-connect-and-react-router)

>   * [Хуки поддерживают статическую типизацию?](#do-hooks-work-with-static-typing)

Reviewer 2: ```suggestion
  * [Поддерживают ли хуки статическую типизацию?](#do-hooks-work-with-static-typing)
```

  * [How does React associate Hook calls with components?](#how-does-react-associate-hook-calls-with-components)
  * [What is the prior art for Hooks?](#what-is-the-prior-art-for-hooks)

## Внедрение хуков {#adoption-strategy}

### В какой версии React появились хуки?{#which-versions-of-react-include-hooks}

Начиная с релиза 16.8.0, React включает в себя стабильную реализацию Хуков для:

* React DOM
* React DOM Server
* React Test Renderer
* React Shallow Renderer

> Обратите внимание, **чтобы Хуки стали доступны, все React-пакеты должны быть версии 16.8.0 или выше**. Хуки не будут работать, если вы, например, забудете обновить React DOM.

Reviewer 2: ```suggestion
Обратите внимание, что **хуки будут доступны, только если все React-пакеты версии 16.8.0 или выше**. Хуки не будут работать, если вы, например, забыли обновить React DOM.
```

Поддержка Хуков в React Native добавится в следующем стабильном релизе.

### Надо ли переписать все мои классовые компоненты?{#do-i-need-to-rewrite-all-my-class-components}

No. There are [no plans](/docs/hooks-intro.html#gradual-adoption-strategy) to remove classes from React -- we all need to keep shipping products and can't afford rewrites. We recommend trying Hooks in new code.

> Нет. [Планов по удалению классов из React нет](/docs/hooks-intro.html#gradual-adoption-strategy) -- нам всем необходимо добавлять новую функциональность, вместо переписывания всего приложения. Мы рекомендуем пробовать Хуки в новом коде.

Reviewer 2: Кажется, так будет понятнее.
```suggestion
Нет. [Мы не собираемся удалять классы из React](/docs/hooks-intro.html#gradual-adoption-strategy). Это новая функциональность, которая не заставляет переписывать всё приложение. Мы рекомендуем пробовать Хуки только в новом коде.
```

### Что я могу сделать с помощью Хуков, что не смог бы с помощью классов? {#what-can-i-do-with-hooks-that-i-couldnt-with-classes}

> Хуки дают отличный способ повторного использования кода между компонентами. ["Создание ваших Хуков"](/docs/hooks-custom.html) показывает, что вы сможете сделать. [Эта статья](https://medium.com/@dan_abramov/making-sense-of-react-hooks-fdbde8803889), написанная одим из главных участников React-команды, описывает новые возможности, разблокированные релизом Хуков.

Reviewer 2: ```suggestion
Хуки дают новый мощный способ переиспользования кода между компонентами. ["Создание ваших хуков"](/docs/hooks-custom.html) даст представление об их возможностях. [Эта статья](https://medium.com/@dan_abramov/making-sense-of-react-hooks-fdbde8803889), написанная одним из участников команды React, описывает новые возможности, которые открывают хуки.
```

### Какая часть моих знаний о React всё еще актуальна? {#how-much-of-my-react-knowledge-stays-relevant}

> Хуки -- это новый способ использование возможностей React, которые вы уже знаете, таких, как состояние, жизненный цикл, контекст и рефы. Хуки не изменили фундаментальную логику React, и ваши знания компонентов, пропсов и потока данных сверху-вниз остаются актуальными.

Reviewer 2: ```suggestion
Хуки -- это новый способ использовать возможности React, которые вы уже знаете: состояние, жизненный цикл, контекст и рефы. Хуки не изменили фундаментальную логику React, и ваши знания компонентов, пропсов и потока данных сверху-вниз остаются актуальными.
```

> Однако, Хуки -- это не самая простая часть React. Если вам не хватает чего-то в этой документации, [дайте нам знать](https://github.com/reactjs/reactjs.org/issues/new), и мы постараемся вам помочь.

Reviewer 2: ```suggestion
Однако, хуки не самая простая часть React. Если вам чего-то не хватает в этой документации, [дайте нам знать](https://github.com/reactjs/reactjs.org/issues/new), и мы постараемся вам помочь.
```

### Что мне использовать: Хуки, классы или оба подхода? {#should-i-use-hooks-classes-or-a-mix-of-both}

> Когда вы будете готовы, мы бы очень хотели, чтобы вы начали использовать Хуки в ваших новых компонентах. Убедитесь, чтобы ваши коллеги ознакомились с документацией и поддержали вас. Мы не рекомендуем сразу перепесивать ваши существующие классы на Хуки, пока вы сами не запланируете переписать их (например, чтобы исправить баг).

Reviewer 2: ```suggestion
Когда вы будете готовы, мы рекомендуем вам начать использовать хуки в ваших новых компонентах. Убедитесь, чтобы каждый в вашей команде знаком с документацией и поддерживает вас. Мы не рекомендуем переписывать ваши существующие классы на хуки, если это не необходимость (например, чтобы исправить баг).
```

Author: @gcor исправил в последем комите, немного улучшив

});

it('can render and update a counter', () => {
  // Тестируем первый рендер и эффект
  act(() => {
    ReactDOM.render(<Counter />, container);
  });
  const button = container.querySelector('button');
  const label = container.querySelector('p');
  expect(label.textContent).toBe('Вы нажали 0 раз');
  expect(document.title).toBe('Вы нажали 0 раз');

  // Тестируем второй рендер и эффект
  act(() => {
    button.dispatchEvent(new MouseEvent('click', {bubbles: true}));
  });
  expect(label.textContent).toBe('Вы нажали 1 раз');
  expect(document.title).toBe('Вы нажали 1 раз');
});
```

Вызовы `act()` также запустят эффекты внутри.

Если вам надо протестировать созданный хук, вы можете создать компонент в ваших тестах и использовать в нём этот хук. После этого вы можете протестировать сам компонент.

Для уменьшения однотипного кода, мы советуем использовать библиотеку [`react-testing-library`](https://git.io/react-testing-library). Она была создана для написания тестов, использующих ваши компоненты, как это делают ваши пользователи.

> ### Что конкретно [правила линтера](https://www.npmjs.com/package/eslint-plugin-react-hooks) проверяют в хуках?{#what-exactly-do-the-lint-rules-enforce}

Reviewer: Аналогично комментарию выше про изменение порядка слов.
```suggestion
### Что конкретно проверяют [правила линтера](https://www.npmjs.com/package/eslint-plugin-react-hooks) в хуках?{#what-exactly-do-the-lint-rules-enforce}
```

> Мы предоставляем [плагин к ESLint](https://www.npmjs.com/package/eslint-plugin-react-hooks), который принуждает следовать [правилам хуков](/docs/hooks-rules.html), чтобы избежать ошибок. Подразумевается, что любая функция, имя который начинается с "`use`" и заглавной буквы далее, является хуком. Мы понимаем, что это предположение не идеальное, и могут случиться ложные срабатывания. Однако, без этого важного соглашения, было бы невозможно заставить хуки работать хорошо, а более длинные имена могли бы помешать людям начать использовать хуки или следовать соглашению.

Reviewer: ```suggestion
Мы предоставляем [плагин к ESLint](https://www.npmjs.com/package/eslint-plugin-react-hooks), который принуждает следовать [правилам хуков](/docs/hooks-rules.html), чтобы избежать ошибок. Подразумевается, что любая функция, имя который начинается с «`use`» и заглавной буквы далее, является хуком. Мы понимаем, что это предположение не идеальное, и могут случиться ложные срабатывания. Однако, без этого важного соглашения, было бы невозможно заставить хуки работать хорошо, а более длинные имена могли бы помешать людям начать использовать хуки или следовать соглашению.
```

> Мы предоставляем [плагин к ESLint](https://www.npmjs.com/package/eslint-plugin-react-hooks), который принуждает следовать [правилам хуков](/docs/hooks-rules.html), чтобы избежать ошибок. Подразумевается, что любая функция, имя который начинается с «`use`» и заглавной буквы далее, является хуком. Мы понимаем, что это предположение не идеальное, и могут случиться ложные срабатывания. Однако, без этого важного соглашения, было бы невозможно заставить хуки работать хорошо, а более длинные имена могли бы помешать людям начать использовать хуки или следовать соглашению.

Reviewer 3: ```suggestion
Мы предоставляем [плагин к ESLint](https://www.npmjs.com/package/eslint-plugin-react-hooks), который принуждает следовать [правилам хуков](/docs/hooks-rules.html), чтобы избежать ошибок. Подразумевается, что любая функция, имя который начинается с «`use`» и заглавной буквы далее, является хуком. Мы понимаем, что это предположение не идеальное, и могут случаться ложные срабатывания. Однако, без этого важного соглашения, было бы невозможно заставить хуки работать хорошо, а более длинные имена могли бы помешать людям начать использовать хуки или следовать соглашению.
```

В частности, правила проверяют, что:

> * Вызовы хуков происходят либо внутри функции с именем в `ВерблюжемРегистре` (подразумевается, что это компонент), либо внутри другой функции `useSomething` (подразумевается, что это созданный пользователем хук).

Reviewer 3: ```suggestion
* Вызовы хуков происходят либо внутри функции с именем, написанном в стиле `PascalCase` (подразумевается, что это компонент), либо внутри другой функции `useSomething` (подразумевается, что это созданный пользователем хук).

```

* Хуки вызываются в одном и том же порядке каждый раз.

> Существует еще несколько предположений, которые могут измениться по мере того, как мы будем регулировать правила и искать баланс между нахождением ошибок и ложными срабатываниями.

Reviewer: ```suggestion
Существует ещё несколько предположений, которые могут измениться по мере того, как мы будем регулировать правила и искать баланс между нахождением ошибок и ложными срабатываниями.
```

## От классов к хукам {#from-classes-to-hooks}

### Как методы жизненного цикла соответствуют хукам? {#how-do-lifecycle-methods-correspond-to-hooks}

> * `constructor`: Функциональному компоненту не нужен конструктор.  Вы можете инициализировать состояние, используя вызов [`useState`](/docs/hooks-reference.html#usestate). Если вычисления состояния трудозатратны, вы можете передать функцию в `useState`.

Reviewer: Может вместо "трудозатратны" использовать "затратны"?
```suggestion
* `constructor`: Функциональному компоненту не нужен конструктор. Вы можете инициализировать состояние, используя вызов [`useState`](/docs/hooks-reference.html#usestate). Если вычисления состояния затратны, вы можете передать функцию в `useState`.
```

* `getDerivedStateFromProps`: Запланировать обновление [вместо рендера](#how-do-i-implement-getderivedstatefromprops).

* `shouldComponentUpdate`: Смотрите объяснение `React.memo` [далее](#how-do-i-implement-shouldcomponentupdate).

* `render`: Это тело функционального компонента.

* `componentDidMount`, `componentDidUpdate`, `componentWillUnmount`: [Хук `useEffect`](/docs/hooks-reference.html#useeffect) заменяет все их комбинации (включая [более](#can-i-skip-an-effect-on-updates) [редкие](#can-i-run-an-effect-only-on-updates) случаи).

* `componentDidCatch` и `getDerivedStateFromError`: В данный момент не существует хуков-аналогов для этих методов, но они будут скоро добавлены.

### Существует что-нибудь наподобие полей экземпляра? {#is-there-something-like-instance-variables}

> Да! Хук [`useRef()`](/docs/hooks-reference.html#useref) может использоваться не только для DOM рефов. Реф – это общий контейнер, а его поле `current` -- изменяемое и может хранить любое значение, как и поле экземпляра класса.

Reviewer 3: ```suggestion
Да! Хук [`useRef()`](/docs/hooks-reference.html#useref) может использоваться не только для DOM-рефов. Реф – это общий контейнер, а его поле `current` -- изменяемое и может хранить любое значение, как и поле экземпляра класса.

```

> Да! Хук [`useRef()`](/docs/hooks-reference.html#useref) может использоваться не только для DOM рефов. Реф - это общий контейнер, а его поле `current` -- изменяемое и может хранить любое значение, как и поле экземпляра класса.

Reviewer: ```suggestion
Да! Хук [`useRef()`](/docs/hooks-reference.html#useref) может использоваться не только для DOM рефов. Реф – это общий контейнер, а его поле `current` -- изменяемое и может хранить любое значение, как и поле экземпляра класса.
```

  // ...
```

> В целом вы можете рассматривать рефы как поля экземпляра класса. Старайтесь избегать установки рефов во время рендера, пока вы не занимаетесь  [ленивой инициализацией](#how-to-create-expensive-objects-lazily), так как это может привести к неожиданному поведению. Вместо этого, зачастую вы будете изменять рефы в обработчиках событий и эффектах.

Reviewer: ```suggestion
В целом вы можете рассматривать рефы как поля экземпляра класса. Старайтесь избегать установки рефов во время рендера, пока вы не занимаетесь [ленивой инициализацией](#how-to-create-expensive-objects-lazily), так как это может привести к неожиданному поведению. Вместо этого, зачастую вы будете изменять рефы в обработчиках событий и эффектах.
```

> ### Сколько переменных состояния я могу использовать - одну или несколько? {#should-i-use-one-or-many-state-variables}

Reviewer: ```suggestion
### Сколько переменных состояния я могу использовать – одну или несколько? {#should-i-use-one-or-many-state-variables}
```

> Если вы привыкли писать классовые компоненты, вы скорее всего захотите вызывать `useState()` один раз для одного изменения и храненить всё состояние в одном объекте. Да, вы можете так сделать, если пожелаете. Взгляните на пример компонента, который следует за движением  курсора. Мы храним позицию и размер этого компонента в локальном состоянии.

Reviewer: ```suggestion
Если вы привыкли писать классовые компоненты, вы скорее всего захотите вызывать `useState()` один раз для одного изменения и хранить всё состояние в одном объекте. Да, вы можете так сделать, если пожелаете. Взгляните на пример компонента, который следует за движением курсора. Мы храним позицию и размер этого компонента в локальном состоянии.
```

}
```

Допустим, мы хотим написать некоторую логику, которая будет изменять значения `left` и `top`, когда пользователь двигает курсор мышки. Обратите внимание, что мы обязаны объединять эти поля с предыдущим объектом состоянии вручную:

```js{4,5}
  // ...
  useEffect(() => {
    function handleWindowMouseMove(e) {
      // Использование "...state" гарантирует, что мы не потеряем поля width и height
      setState(state => ({ ...state, left: e.pageX, top: e.pageY }));
    }

>     // Примичание: эта реализация немного упрощена

Reviewer 3: ```suggestion
    // Примечание: эта реализация немного упрощена
```

    // ...
```

> В разделении состояния на независимые переменные есть еще одно преимущество. Это поможет легко вынести некоторую логику в отдельный хук в будущем, например:

Reviewer: ```suggestion
В разделении состояния на независимые переменные есть ещё одно преимущество. Это поможет легко вынести некоторую логику в отдельный хук в будущем, например:
```

}
```

Обратите внимание, как мы смогли вынести вызов `useState`, изменяющий `position`, и соответствующих эффект в отдельный хук, практически не меняя их код. Если бы всё состояние хранилось в одном объекте, сделать это перемещение было бы на порядок сложнее.

На самом деле, оба варианта имеют право на жизнь. Однако, компоненты будут более читаемыми, если вы найдёте баланс между двумя подходами и будете группировать связанные друг с другом данные. Если работа с состоянием становится сложной, мы советуем [использовать редюсер](/docs/hooks-reference.html#usereducer) или выносить логику в отдельные хуки.

### Могу ли я использовать эффект только на обновлениях компонента? {#can-i-run-an-effect-only-on-updates}

Это довольно редкий случай. Вы можете [использовать изменяемый реф](#is-there-something-like-instance-variables), чтобы вручную хранить логическое значение, показывающее, прошел ли первый рендер. В свою очередь эффект может опираться на это значение. Если вы замечаете, что эта логика нужна вам часто, вы можете вынести её в отдельный хук.

> ### Как получить прошлые пропсы или состояние? {#how-to-get-the-previous-props-or-state}

Reviewer 3: ```suggestion
### Как получить предыдущие пропсы или состояние? {#how-to-get-the-previous-props-or-state}
```

  // ...
```

> Возможно, в будущем React добавит хук `usePrevious` в свой API, так как это довольно распространенный случай в разработке.

Reviewer: ```suggestion
Возможно, в будущем React добавит хук `usePrevious` в свой API, так как это довольно распространённый случай в разработке.
```

Также смотрите [рекомендованный паттерн для полученного состояния](#how-do-i-implement-getderivedstatefromprops).

### Как я могу реализовать `getDerivedStateFromProps`? {#how-do-i-implement-getderivedstatefromprops}

> Не смотря на то, что вам скорее всего [это не нужно](/blog/2018/06/07/you-probably-dont-need-derived-state.html), в редких случаях (например, реализация компонента `<Transition>`), вы можете изменять состояние прямо во время рендера. React моментально сделает повторный рендер компонента с обновлённым состоянием сразу после первого рендера, и это не будет затратно.

Reviewer 3: ```suggestion
Несмотря на то, что вам скорее всего [это не нужно](/blog/2018/06/07/you-probably-dont-need-derived-state.html), в редких случаях (например, реализация компонента `<Transition>`), вы можете изменять состояние прямо во время рендера. React моментально сделает повторный рендер компонента с обновлённым состоянием сразу после первого рендера, и это не будет затратно.
```

}
```

> Сперва это может показаться странным, но изменение во время рендера - это именно то, чем всегда концептуально являлся метод `getDerivedStateFromProps`.

Reviewer: ```suggestion
Сперва это может показаться странным, но изменение во время рендера – это именно то, чем всегда концептуально являлся метод `getDerivedStateFromProps`.
```

### Существует что-нибудь наподобие forceUpdate? {#is-there-something-like-forceupdate}

> Оба хука `useState` и `useReducer` [игнорируют апдейты](/docs/hooks-reference.html#bailing-out-of-a-state-update), если следующее значение равно предыдущему. Мутирование состояния и вызов `setState` не вызовет повторного рендера.

Reviewer: ```suggestion
Оба хука `useState` и `useReducer` [игнорируют обновления](/docs/hooks-reference.html#bailing-out-of-a-state-update), если следующее значение равно предыдущему. Мутация состояния и вызов `setState` не вызовет повторного рендера.
```

> Обычно, вы не должны мутировать локальное состояние в React. Однако, в качестве трюка, вы можете использовать увеличивающийся счётчик, чтобы заставить компонент повторно рендериться, если состояние не изменилось:

Reviewer 3: ```suggestion
Обычно, вы не должны мутировать внутреннее состояние в React. Однако, в качестве трюка, вы можете использовать увеличивающийся счётчик, чтобы заставить компонент повторно рендериться, если состояние не изменилось:
```

> Обычно, вы не должны мутировать локальное состояние в React. Однако, в качестве трюка, вы можете использовать увеличивающийся счетчик, чтобы заставлять компонент повторно рендериться, если состояние не изменилось:

Reviewer: ```suggestion
Обычно, вы не должны мутировать локальное состояние в React. Однако, в качестве трюка, вы можете использовать увеличивающийся счётчик, чтобы заставить компонент повторно рендериться, если состояние не изменилось:
```

### Как я могу реализовать `shouldComponentUpdate`? {#how-do-i-implement-shouldcomponentupdate}

Вы можете обернуть функциональный компонент в вызов `React.memo` для поверхностного сравнения его пропсов:

```js
const Button = React.memo((props) => {
  // ваш компонент
});
```

Это функция не является хуком, так как она не ведёт себя как хук. `React.memo` аналогична `PureComponent`, но она сравнивает только пропсы. (Вы также можете добавить второй аргумент, чтобы определить свою функцию сравнения, которая примет старые и новые пропсы. Если эта функция вернёт `true`, обновление будет пропущено.)

`React.memo` не сравнивает состояние, потому что не существует единого объекта, который можно сравнить. Но вы можете также сделать дочерние компоненты чистыми или даже [оптимизировать определённый дочерний компонент используя хук `useMemo`](/docs/hooks-faq.html#how-to-memoize-calculations).

> ### Как закэшировать вычисления? {#how-to-memoize-calculations}

Reviewer 3: ```suggestion
### Как закешировать вычисления? {#how-to-memoize-calculations}
```

Хук [`useMemo`](/docs/hooks-reference.html#usememo) позволяет вам закэшировать вычисления между несколькими рендерами, путём запоминания прошлого результата:

```js
const memoizedValue = useMemo(() => computeExpensiveValue(a, b), [a, b]);
```

Этот код вызовет `computeExpensiveValue(a, b)`. Но если аргументы `[a, b]` не изменились с прошлого рендера, `useMemo` пропустит повторный вызов и повторно использует значения, которые он вернул в прошлый раз.

Помните, что функция, передаваемая в `useMemo`, выполняется во время рендера. Не стоит делать в ней что-то, что вы обычно не делаете во время рендера. Например, побочные эффекты выполняются в хуке `useEffect`, а не в `useMemo`.

> **Рассматриваете `useMemo`, как оптимизацию производительности, но не стоит полагаться на неё на 100%.** В будущем React сможет `забыть` прошлые закэшированные значения и провести перерасчет на следующем рендере, например, для освобождения памяти, занятой компонентами вне экрана. Пишите ваш код так, чтобы он мог работать без `useMemo`, и только тогда добавляйте оптимизацию производительности. (Для редких случаев, когда значение **не должно никогда** пересчитываться, вы можете [лениво инициализировать](#how-to-create-expensive-objects-lazily) реф.)

Reviewer: ```suggestion
**Рассматриваете `useMemo`, как оптимизацию производительности, но не стоит полагаться на неё на 100%.** В будущем React сможет `забыть` прошлые закэшированные значения и провести перерасчёт на следующем рендере, например, для освобождения памяти, занятой компонентами вне экрана. Пишите ваш код так, чтобы он мог работать без `useMemo`, и только тогда добавляйте оптимизацию производительности. (Для редких случаев, когда значение **не должно никогда** пересчитываться, вы можете [лениво инициализировать](#how-to-create-expensive-objects-lazily) реф.)
```

> **Рассматриваете `useMemo`, как оптимизацию производительности, но не стоит полагаться на неё на 100%.** В будущем React сможет `забыть` прошлые закэшированные значения и провести перерасчёт на следующем рендере, например, для освобождения памяти, занятой компонентами вне экрана. Пишите ваш код так, чтобы он мог работать без `useMemo`, и только тогда добавляйте оптимизацию производительности. (Для редких случаев, когда значение **не должно никогда** пересчитываться, вы можете [лениво инициализировать](#how-to-create-expensive-objects-lazily) реф.)

Reviewer 3: ```suggestion
**Рассматриваете `useMemo`, как оптимизацию производительности, но не стоит полагаться на неё на 100%.** В будущем React сможет `забыть` прошлые закешированные значения и провести перерасчёт на следующем рендере, например, для освобождения памяти, занятой компонентами вне экрана. Пишите ваш код так, чтобы он мог работать без `useMemo`, и только тогда добавляйте оптимизацию производительности. (Для редких случаев, когда значение **не должно никогда** пересчитываться, вы можете [лениво инициализировать](#how-to-create-expensive-objects-lazily) реф.)
```

}
```

Обратите внимание, что этот подход не будет работать внутри циклов, так как вызовы хуков [не могут](/docs/hooks-rules.html) быть помещены внутрь циклов. Но вы можете создать отдельный компонент для элемента списка, и вызвать `useMemo` там.

### Как лениво создавать большие объекты? {#how-to-create-expensive-objects-lazily}

> Хук `useMemo` позволяет вам [закэшировать ресурсозатратные вычисления](#how-to-memoize-calculations), если входные данные не изменились. Однако, этот способ **не гарантирует**, что повторных вычислений не случится. Иногда вам нужно удостовериться, что объект будет создан только один раз.

Reviewer 3: ```suggestion
Хук `useMemo` позволяет вам [закешировать ресурсозатратные вычисления](#how-to-memoize-calculations), если входные данные не изменились. Однако, этот способ **не гарантирует**, что повторных вычислений не случится. Иногда вам нужно удостовериться, что объект будет создан только один раз.
```

> **Первый частый вариант использования - создание изначального состояния ресурсозатратно:**

Reviewer: Аналогично комментарию выше про замену "трудозатратно" на "затратно" Может здесь тоже заменить?
```suggestion
**Первый частый вариант использования – создание изначального состояния затратно:**
```

```js
function Table(props) {
  // ⚠️ createRows() вызывается на каждом рендере
  const [rows, setRows] = useState(createRows(props.count));
  // ...
}
```

> Чтобы повторно не создовать начальное состояние (которое по факту игнорируется), мы можем передать **функцию** в `useState`:

Reviewer: ```suggestion
Чтобы повторно не создавать начальное состояние (которое по факту игнорируется), мы можем передать **функцию** в `useState`:
```

    return newObserver;
  }

>   // Когда он вам понадобится, вызовете getObserver()

Reviewer: ```suggestion
  // Когда он вам понадобится, вызовите getObserver()
```

  // ...
}
```

> Этот подход позволяет избежать создания больших объектов на всех этапах, кроме первого рендера, кода это действительно необходимо. Если вы используете Flow или TypeScript, вы можете также установить для `getObserver()` непустой тип для уверенности.

Reviewer: ```suggestion
Этот подход позволяет избежать создания больших объектов на всех этапах, кроме первого рендера, когда это действительно необходимо. Если вы используете Flow или TypeScript, вы можете также установить для `getObserver()` непустой тип для уверенности.
```

### Являются ли хуки медленными из-за создания функций на каждом рендере? {#are-hooks-slow-because-of-creating-functions-in-render}

Нет. В современных браузерах чистая производительность замыканий не сильно отличается от классов, кроме особых случаев.

Также учитывайте, что реализация хуков более эффективная по нескольким причинам:

> * Хуки не требуют больших затрат, которые неизбежны в классах. Например,  создание экземпляра класса и связывание обработчиков событий в конструкторе.

Reviewer: ```suggestion
* Хуки не требуют больших затрат, которые неизбежны в классах. Например, создание экземпляра класса и связывание обработчиков событий в конструкторе.
```

}
```

> Этот вариант более удобный, как с точки зрения поддержки кода (не нужно добавлять пропсы для новых колбэков), так и в разрезе избежания общих проблем с колбэками. Передача функции `dispatch` по такому принципу - это рекомендуемый подход для вызова колбэков и избежания нежелательных обновлений в глубине дерева компонентов.

Reviewer: ```suggestion
Этот вариант более удобный, как с точки зрения поддержки кода (не нужно добавлять пропсы для новых колбэков), так и в разрезе избежания общих проблем с колбэками. Передача функции `dispatch` по такому принципу – это рекомендуемый подход для вызова колбэков и избежания нежелательных обновлений в глубине дерева компонентов.
```

Обратите внимание, что вы также можете передавать *состояние* приложения вниз через пропсы (более явно) или через контекст (более удобно для глубокого дерева). Если вы используете контекст также для передачи состояния, используете два разных контекста, так как функция `dispatch` никогда не изменяется. Поэтому компоненты, использующие только эту функцию, не будут нуждаться в повторном рендере, пока они не начнут использовать второй контекст с состоянием.

### Как получить часто изменяемое значение из хука `useCallback`? {#how-to-read-an-often-changing-value-from-usecallback}

> Примечание
>
> Мы рекомендуем [передавать `dispatch` вниз через контекст](#how-to-avoid-passing-callbacks-down) вместо передачи отдельных колбэков через пропсы. Следующий подход приведён тут только для полноты картины в качестве трюка.
>

> > Также обратите внимание, что этот подход может вызвать проблемы в [параллельном режиме](/blog/2018/03/27/update-on-async-rendering.html). Мы планируем предоставить более удобные альтернативы в будущем, но самый безопасный вариант сейчас -- всегда инвалидировать колбэк, если значение, на которое он опирается, поменялось.

Reviewer 3: ```suggestion
> Также обратите внимание, что этот подход может вызвать проблемы в [параллельном режиме](/blog/2018/03/27/update-on-async-rendering.html). Мы планируем предоставить более удобные альтернативы в будущем, но самый безопасный вариант сейчас -- всегда отменять колбэк, если значение, на которое он опирается, поменялось.
```

> В некоторых редких случаях, вам может понадобится закэшировать колбэк с помощью хука [`useCallback`](/docs/hooks-reference.html#usecallback), но кэширование не будет работать хорошо, потому что внутренняя функция будет пересоздоваться слишком часто. Если функция, которую вы кэшируете, является обработчиком событий и не используется во время рендера, вы можете использовать [реф в качестве поля экземпляра](#is-there-something-like-instance-variables), и вручную сохранить туда последнее значение:

Reviewer 3: ```suggestion
В некоторых редких случаях, вам может понадобится закешировать колбэк с помощью хука [`useCallback`](/docs/hooks-reference.html#usecallback), но кеширование не будет работать хорошо, потому что внутренняя функция будет пересоздоваться слишком часто. Если функция, которую вы кешируете, является обработчиком событий и не используется во время рендера, вы можете использовать [реф в качестве поля экземпляра](#is-there-something-like-instance-variables), и вручную сохранить туда последнее значение:
```

}
```

Это крайне тяжелый подход, но он показывает, что вы можете сделать эту хитрую оптимизацию, если это действительно вам необходимо. В любом случае, советуем вынести эту логику в отдельный хук:

```js{4,16}
function Form() {
  const [text, updateText] = useState('');

>   // Будет закэшировано, даже если значение `text` измениться:

Reviewer 3: ```suggestion
  // Будет закешировано, даже если значение `text` измениться:
```

}
```

В любом случае мы **не советуем использовать этот подход** и показываем его тут только для полноты картины. Вместо этого, старайтесь [избегать передачи колбэков вниз глубоко в дерево](#how-to-avoid-passing-callbacks-down).

## Под капотом {#under-the-hood}

### Как React связывает вызовы хуков с компонентом? {#how-does-react-associate-hook-calls-with-components}

React следит за тем, какой компонент рендерится в данный момент. Благодаря [правилам хуков](/docs/hooks-rules.html), мы знаем, что хуки вызываются только из React-компонентов (отдельные хуки, написанные вами, также вызываются только из React-компонентов).

> Существует внутренний список `ячеек памяти`, связанных с каждым компонентом. Они являются JavaScript-объектами, в которых мы можем хранить некоторые данные. Когда вы вызывается некий хук, например `useState()`, он читает значение текущей ячейки (или инициализирует её во время первого рендера) и двигает указатель на следующую. Таким способом каждый вызов `useState()` получит своё независимое состояние.

Reviewer: ```suggestion
Существует внутренний список `ячеек памяти`, связанных с каждым компонентом. Они являются JavaScript-объектами, в которых мы можем хранить некоторые данные. Когда вызывается некий хук, например `useState()`, он читает значение текущей ячейки (или инициализирует её во время первого рендера) и двигает указатель на следующую. Таким способом каждый вызов `useState()` получит своё независимое состояние.
```

> ### Что послужило прообразом Хуков? {#what-is-the-prior-art-for-hooks}

Reviewer 3: ```suggestion
### Что послужило прообразом хуков? {#what-is-the-prior-art-for-hooks}
```

Хуки были основаны на следующих идеях и экспериментах:

* Наши старые эксперименты с функциональным API в репозитории [react-future](https://github.com/reactjs/react-future/tree/master/07%20-%20Returning%20State).

> * Эксперименты с рендер-пропс API от React-сообщества, включая [Reactions-компонент](https://github.com/reactions/component) от [Ryan Florence](https://github.com/ryanflorence).

Reviewer 3: ```suggestion
* Эксперименты с API рендер-пропс от React-сообщества, включая [компонент Reactions](https://github.com/reactions/component) от [Райна Флоренса (Ryan Florence)](https://github.com/ryanflorence).
```

* [Dominic Gannaway](https://github.com/trueadm) и его предложение [ключевого слова `adopt`](https://gist.github.com/trueadm/17beb64288e30192f3aa29cad0218067), в качестве синтаксического сахара для паттерна рендер-пропс.
* Переменные и ячейки состояния в [DisplayScript](http://displayscript.org/introduction.html).
* [Компоненты-редюсеры](https://reasonml.github.io/reason-react/docs/en/state-actions-reducer.html) в ReasonReact.
* [Подписки](http://reactivex.io/rxjs/class/es6/Subscription.js~Subscription.html) в Rx.
* [Алгебраические эффекты](https://github.com/ocamllabs/ocaml-effects-tutorial#2-effectful-computations-in-a-pure-setting) в Multicore OCaml.

> [Sebastian Markbåge](https://github.com/sebmarkbage) предложил изначальный дизайн хуков, который позднее улучшили [Andrew Clark](https://github.com/acdlite), [Sophie Alpert](https://github.com/sophiebits), [Dominic Gannaway](https://github.com/trueadm) и другие участники команды React.

Reviewer 3: ```suggestion
[Себастьян Маркбаге (Sebastian Markbåge)](https://github.com/sebmarkbage) предложил изначальный дизайн хуков, который позднее улучшили [Эндрю Кларк (Andrew Clark)](https://github.com/acdlite), [Софи Алперт (Sophie Alpert)](https://github.com/sophiebits), [Доминик Ганнауэй (Dominic Gannaway)](https://github.com/trueadm) и другие участники команды React.
```

}
```

В любом случае мы **не советуем использовать этот подход** и показываем его тут только для полноты картины. Вместо этого, старайтесь [избегать передачи колбэков вниз глубоко в дерево](#how-to-avoid-passing-callbacks-down).

## Под капотом {#under-the-hood}

### Как React связывает вызовы хуков с компонентом? {#how-does-react-associate-hook-calls-with-components}

> React следит за тем, какой компонент рендерится в данный момент. Благодаря [правилам хуков](/docs/hooks-rules.html), мы знаем, что хуки вызываются только из React-компонентов (отдельные хуки, написанные вами, также вызываются только из React-компонентов).

Reviewer 4: ```suggestion
React следит за тем, какой компонент рендерится в данный момент. Благодаря [правилам хуков](/docs/hooks-rules.html), мы знаем, что хуки вызываются только из React-компонентов (пользовательские хуки также вызываются только из React-компонентов).
```
